rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -836,148 +823,282 @@ rmw_ret_t configure_qos_for_security(
-       entirely in line with other entity types).  In this case, we don't have to do this
-       (though there would be no harm in it either).
+  rmw_subscription_options_t subscription_options = rmw_get_default_subscription_options();
+  subscription_options.ignore_local_publications = true;
+  // FIXME: keyed topics => KEEP_LAST and depth 1.
+  pubsub_qos.history = RMW_QOS_POLICY_HISTORY_KEEP_ALL;
+  std::unique_ptr<rmw_subscription_t, std::function<void(rmw_subscription_t *)>>
+  subscription(
+    create_subscription(
+      impl->ppant, impl->ppant,
+      rosidl_typesupport_cpp::get_message_type_support_handle<ParticipantEntitiesInfo>(),
+      "ros_discovery_info",
+      &pubsub_qos,
+      &subscription_options),
+    [&](rmw_subscription_t * sub) {
+      if (RMW_RET_OK != destroy_subscription(sub)) {
+        RCUTILS_SAFE_FWRITE_TO_STDERR(
+          "Failed to destroy subscription after function: '"
+          RCUTILS_STRINGIFY(__function__) "' failed.\n");
+      }
+    });
+  if (subscription == nullptr) {
+    return RMW_RET_ERROR;
+  }
 
-       Note that "domains_lock" is held all this time. */
-    gcdds.domains[did].domain_handle = dds_get_parent(pp);
+  std::unique_ptr<rmw_guard_condition_t, std::function<void(rmw_guard_condition_t *)>>
+  graph_guard_condition(
+    create_guard_condition(impl.get()),
+    [&](rmw_guard_condition_t * gc) {
+      if (RMW_RET_OK != destroy_guard_condition(gc)) {
+        RMW_SAFE_FWRITE_TO_STDERR(
+          "Failed to destroy guard condition after function: '"
+          RCUTILS_STRINGIFY(__function__) "' failed.\n");
+      }
+    });
+  if (!graph_guard_condition) {
+    return RMW_RET_BAD_ALLOC;
+  }
+
+  impl->common.graph_cache.set_on_change_callback(
+    [guard_condition = graph_guard_condition.get()]() {
+      static_cast<void>(rmw_trigger_guard_condition(guard_condition));
+    });
+
+  impl->common.pub = publisher.get();
+  impl->common.sub = subscription.get();
+  impl->common.graph_guard_condition = graph_guard_condition.get();
+  get_entity_gid(impl->ppant, impl->common.gid);
+  impl->common.graph_cache.add_participant(impl->common.gid, context->options.security_context);
+  impl->common.thread_is_running.store(false);
+  impl->common.listener_thread_gc = nullptr;
+
+  context->impl = impl.get();
+
+  // FIXME: not sure if I could get away with using a bunch of listeners
+  // - Locally published samples shouldn't make it to the reader, so there shouldn't be a deadlock
+  //   caused by the graph cache's mutex already having been locked by (e.g.) rmw_create_node.
+  // - Whatever the graph cache implementation does, it shouldn't involve much more than local state
+  //   updates and triggering a guard condition, and so that should be safe.
+  // - There might be a significant cost to calling the graph cache function to update its internal
+  //   state, and having that occur on the thread receiving data from the network may not be wise
```

Comment:
Yeah, the last point is the main reason I've used a separate thread.
`rmw_dds_common` can probably be largely improved (and the way ros discovery data is shared too).